opm/simulators/linalg/PressureSolverPolicy.hpp

```diff
@@ -0,0 +1,128 @@
+/*
+  Copyright 2019 SINTEF Digital, Mathematics and Cybernetics.
+
+  This file is part of the Open Porous Media project (OPM).
+
+  OPM is free software: you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  OPM is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#ifndef OPM_PRESSURE_SOLVER_POLICY_HEADER_INCLUDED
+#define OPM_PRESSURE_SOLVER_POLICY_HEADER_INCLUDED
+
+#include <opm/simulators/linalg/PressureTransferPolicy.hpp>
+
+#include <boost/property_tree/ptree.hpp>
+
+#include <dune/istl/solver.hh>
+
+namespace Dune
+{
+namespace Amg
+{
+    namespace pt = boost::property_tree;
+
+    template <class OperatorType, class Solver, class LevelTransferPolicy>
+    class PressureSolverPolicy
+    {
+    public:
+        /** @brief The type of the linear operator used. */
+        using Operator = OperatorType;
+        /**
+         * @brief Constructs the coarse solver policy.
+         * @param prm Parameter tree specifying the solver details.
+         */
+        explicit PressureSolverPolicy(const pt::ptree prm)
+            : prm_(prm)
+        {
+        }
+
+    private:
+        using X = typename Operator::range_type;
+        /**
+         * @brief A wrapper that makes an inverse operator out of AMG.
+         *
+         * The operator will use one step of AMG to approximately solve
+         * the coarse level system.
+         */
+        struct PressureInverseOperator : public Dune::InverseOperator<X, X> {
+            template <class Comm>
+            PressureInverseOperator(Operator& op, const boost::property_tree::ptree& prm, const Comm& comm)
+                : linsolver_()
+            {
+                if (op.category() == Dune::SolverCategory::overlapping) {
+                    linsolver_.reset(new Solver(prm, op.getmat(), comm));
+                } else {
+                    linsolver_.reset(new Solver(prm, op.getmat()));
+                }
+            }
+
+            Dune::SolverCategory::Category category() const override
+            {
+                return Dune::SolverCategory::sequential;
```

Comment:
Wouldn't it be handy to export the real category?

Reply:
Yes, this is a bug.